Sources/FoundationInternationalization/Formatting/Number/BinaryInteger+FormatStyle.swift

```diff
@@ -43,250 +43,187 @@ extension BinaryInteger {
-                }
-            }
-
-            preconditionFailure("\(type(of: self)) \(self) compared as not zero yet all its words are zero.")
-        } else { // `self` is negative.
-            // Perform two's complement one word at a time, keeping track of the index of the highest set bit seen so far.  After enumerating all the words, return that index plus one (index to count conversion).
-
-            var carryingOne = true // Covers both the initial +1 (as part of two's complement) and overflow between words.
-            var indexOfHighestSetBitSeenSoFar = 0
-
-            for (i, word) in words.enumerated() {
-                var positiveWord = ~word
-
-                if carryingOne {
-                    (positiveWord, carryingOne) = positiveWord.addingReportingOverflow(1)
-                }
-
-                let fullWidth = type(of: word).bitWidth
-
-                if carryingOne {
-                    indexOfHighestSetBitSeenSoFar += fullWidth
-                } else if .zero != positiveWord {
-                    indexOfHighestSetBitSeenSoFar = (fullWidth * i) + (fullWidth - positiveWord.leadingZeroBitCount) - 1
-                }
-            }
-
-            return indexOfHighestSetBitSeenSoFar + 1
+    let radix: (exponent: Int, power: UInt) = maxDecimalExponentAndPowerForUnsignedIntegerWord()
+    let capacity = maxDecimalDigitCountForUnsignedInteger(bitWidth: words.count * UInt.bitWidth) + (isLessThanZero ? 1 : 0)
+    var ascii = ContiguousArray(repeating: UInt8(ascii: "0"), count: capacity) // Set initial ASCII zeros (see later steps).
+    
+    var wordsIndex = words.endIndex
+    var writeIndex = ascii.endIndex
+    var asciiIndex = ascii.endIndex
+    
+    ascii.withUnsafeMutableBufferPointer { ascii in
+        repeat {
+            // Mutating division prevents unnecessary big integer allocations.
+            let remainder = formQuotientWithRemainderForUnsignedInteger(words: words[..<wordsIndex], dividingBy: radix.power)
+            // Trim the quotient's most significant zeros for flexible-width performance and to end the loop.
+            wordsIndex  = words[..<wordsIndex].reversed().drop(while:{ $0 == 0 as UInt }).startIndex.base
+            // The remainder's numeric string representation is written to the array's trailing edge, up to the ASCII index.
+            writeIndex  = formTrailingNumericStringRepresentationForUnsignedInteger(word: remainder, into: ascii[..<asciiIndex])
+            // Because the array is pre-filled with ASCII zeros, we can skip ahead; only the final loop's zeros are trimmed.
+            asciiIndex -= radix.exponent
+            // The entire magnitude has been encoded when the formed quotient is empty.
+        }   while wordsIndex > words.startIndex
+        
+        // Jump to the most significant digit.
+        asciiIndex = writeIndex
+        
+        // Add a minus sign to negative values.
+        if isLessThanZero {
+            ascii.formIndex(before: &asciiIndex)
```

Comment:
Sorry, I didn't see your question earlier. The `capacity` is given by the following expression so please have a look at [number-of-decimal-digits-in-a-binary-integer](https://www.exploringbinary.com/number-of-decimal-digits-in-a-binary-integer). 

```swift
let capacity = maxDecimalDigitCountForUnsignedInteger(bitWidth: words.count * UInt.bitWidth) + (isLessThanZero ? 1 : 0)
```

Since `capacity` is only an upper bound, I've used the fullest possible width as an uncomplicated approximaton (it can be trimmed a bit). In the loop, we then divide by `10` repeatedly and write each digit back-to-front. The loop is on `repeat`, so we write at least one. The minimum capacity is `1` when `words.count == 0`, and `1` is added when `isLessThanZero == true`.

```
let isLessThanZero = isSigned && Int(bitPattern: words.last ?? 0) < 0 // tests the last bit when isSigned == true
```

#### Important edge case thing-y

The loop branchlessly decrements `asciiIndex` by `radix.exponent` (the digits written plus zeros for padding) each iteration, so it might be negative at the end of the loop. This is one reason why we set it to `writeIndex` afterwards, the other is trimmings the last iteration's padding. This assumes `Index == Int` (alternatives exist).

Reply:
For what it's worth, no index goes out of bounds or assumes typing in v2. I've also replaced all assertions with preconditions since the difference in performance was insignificant.